File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head.next:
            return 
        # locate the first part and the second part
        # we are going to do this using fast and slow pointers:
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        # reverse the second part
        second = slow.next
        sec_next = second.next
        if sec_next:
            tmp = sec_next.next
        else:
            tmp = None
        second.next = None 
        while sec_next:
            sec_next.next = second
            second = sec_next
            sec_next = tmp
            if tmp:
                tmp = tmp.next
        
        

        # merge the two parts
        slow.next = None
        sec = second
        cur = head
        tmp = cur.next
        while sec:
            cur.next = sec
            cur = tmp
            if tmp:
                tmp = tmp.next
            tmp_s = sec.next
            sec.next = cur
            sec = tmp_s
        




        
        

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == nullptr|| head->next == nullptr)
        return false;
    
    ListNode* fast = head->next->next;
    ListNode* slow = head->next;

    while(fast && slow != fast){
        slow = slow -> next;
        fast = fast->next;
        if(!fast) break;
        fast = fast->next; 
    }

    return fast == nullptr ? false: true;
    }
};
